File: packages/sdk-python/src/agentops/prompts/registry.py
"""Prompt registry for version-controlled prompt management."""

from __future__ import annotations

import re
import time
import uuid
from typing import Any

from .types import PromptTemplate, PromptVersion


class PromptRegistry:
    """Version-controlled storage and management of prompt templates."""

    def __init__(self) -> None:
        self._templates: dict[str, PromptTemplate] = {}
        self._versions: dict[str, list[PromptVersion]] = {}

    def register(
        self,
        name: str,
        template: str,
        *,
        id: str | None = None,
        description: str | None = None,
        tags: list[str] | None = None,
        target_model: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> PromptTemplate:
        """Register a new prompt template."""
        template_id = id or f"prompt_{uuid.uuid4().hex[:12]}"
        variables = self._extract_variables(template)
        timestamp = int(time.time() * 1000)

        prompt_template = PromptTemplate(
            id=template_id,
            name=name,
            template=template,
            variables=variables,
            version="1.0.0",
            description=description,
            tags=tags or [],
            target_model=target_model,
            created_at=timestamp,
            updated_at=timestamp,
            metadata=metadata or {},
        )

        self._templates[template_id] = prompt_template
        self._versions[template_id] = [
            PromptVersion(
                version="1.0.0",
                template=template,
                created_at=timestamp,
            )
        ]

        return prompt_template

    def get(self, id: str) -> PromptTemplate | None:
        """Get a prompt template by ID."""
        return self._templates.get(id)

    def get_by_name(self, name: str) -> PromptTemplate | None:
        """Get a prompt template by name."""
        for template in self._templates.values():
            if template.name == name:
                return template
        return None

    def list(
        self,
        *,
        tags: list[str] | None = None,
        target_model: str | None = None,
    ) -> list[PromptTemplate]:
        """List all templates with optional filtering."""
        templates = list(self._templates.values())

        if tags:
            templates = [
                t for t in templates
                if any(tag in t.tags for tag in tags)
            ]

        if target_model:
            templates = [
                t for t in templates
                if t.target_model == target_model
            ]

        return templates

    def update(
        self,
        id: str,
        template: str,
        change_description: str | None = None,
    ) -> PromptTemplate | None:
        """Update a prompt template (creates new version)."""
        existing = self._templates.get(id)
        if not existing:
            return None

        new_version = self._increment_version(existing.version)
        timestamp = int(time.time() * 1000)

        updated = PromptTemplate(
            id=existing.id,
            name=existing.name,
            template=template,
            variables=self._extract_variables(template),
            version=new_version,
            description=existing.description,
            tags=existing.tags,
            target_model=existing.target_model,
            created_at=existing.created_at,
            updated_at=timestamp,
            metadata=existing.metadata,
        )

        self._templates[id] = updated

        if id not in self._versions:
            self._versions[id] = []
        self._versions[id].append(
            PromptVersion(
                version=new_version,
                template=template,
                created_at=timestamp,
                change_description=change_description,
            )
        )

        return updated

    def get_version_history(self, id: str) -> list[PromptVersion]:
        """Get version history for a template."""
        return self._versions.get(id, [])

    def get_version(self, id: str, version: str) -> PromptVersion | None:
        """Get a specific version of a template."""
        history = self._versions.get(id, [])
        for v in history:
            if v.version == version:
                return v
        return None

    def rollback(self, id: str, version: str) -> PromptTemplate | None:
        """Rollback to a previous version."""
        target_version = self.get_version(id, version)
        if not target_version:
            return None

        return self.update(id, target_version.template, f"Rollback to version {version}")

    def delete(self, id: str) -> bool:
        """Delete a template."""
        deleted = id in self._templates
        self._templates.pop(id, None)
        self._versions.pop(id, None)
        return deleted

    def render(
        self,
        id: str,
        variables: dict[str, str],
    ) -> str | None:
        """Render a template with variables."""
        template = self._templates.get(id)
        if not template:
            return None

        return self.render_template(template.template, variables)

    def render_template(
        self,
        template: str,
        variables: dict[str, str],
    ) -> str:
        """Render a template string directly."""
        result = template
        for key, value in variables.items():
            pattern = r"\{\{\s*" + re.escape(key) + r"\s*\}\}"
            result = re.sub(pattern, value, result)
        return result

    def diff(
        self,
        id: str,
        version1: str,
        version2: str,
    ) -> dict[str, list[str]] | None:
        """Compare two versions of a template."""
        v1 = self.get_version(id, version1)
        v2 = self.get_version(id, version2)

        if not v1 or not v2:
            return None

        lines1 = v1.template.split("\n")
        lines2 = v2.template.split("\n")
        set1 = set(lines1)
        set2 = set(lines2)

        return {
            "added": [line for line in lines2 if line not in set1],
            "removed": [line for line in lines1 if line not in set2],
            "unchanged": [line for line in lines1 if line in set2],
        }

    def export_all(self) -> dict[str, PromptTemplate]:
        """Export all templates."""
        return dict(self._templates)

    def import_all(self, templates: dict[str, PromptTemplate]) -> None:
        """Import templates."""
        for id, template in templates.items():
            self._templates[id] = template
            self._versions[id] = [
                PromptVersion(
                    version=template.version,
                    template=template.template,
                    created_at=template.created_at,
                )
            ]

    def _extract_variables(self, template: str) -> list[str]:
        pattern = r"\{\{\s*(\w+)\s*\}\}"
        return list(set(re.findall(pattern, template)))

    def _increment_version(self, version: str) -> str:
        parts = [int(x) for x in version.split(".")]
        parts[-1] += 1
        return ".".join(str(x) for x in parts)

File: packages/sdk-ts/src/prompts/registry.ts
/**
 * AgentOps SDK - Prompt Registry
 *
 * Version-controlled storage and management of prompt templates.
 */

import type { PromptTemplate, PromptVersion } from "./types.js";
import { now } from "../utils.js";
import { nanoid } from "nanoid";

export class PromptRegistry {
  private templates: Map<string, PromptTemplate> = new Map();
  private versions: Map<string, PromptVersion[]> = new Map();

  /**
   * Register a new prompt template
   */
  register(
    name: string,
    template: string,
    options?: {
      id?: string;
      description?: string;
      tags?: string[];
      targetModel?: string;
      metadata?: Record<string, unknown>;
    },
  ): PromptTemplate {
    const id = options?.id ?? `prompt_${nanoid(12)}`;
    const variables = this.extractVariables(template);
    const timestamp = now();

    const promptTemplate: PromptTemplate = {
      id,
      name,
      template,
      variables,
      version: "1.0.0",
      description: options?.description,
      tags: options?.tags,
      targetModel: options?.targetModel,
      createdAt: timestamp,
      updatedAt: timestamp,
      metadata: options?.metadata,
    };

    this.templates.set(id, promptTemplate);
    this.versions.set(id, [
      {
        version: "1.0.0",
        template,
        createdAt: timestamp,
      },
    ]);

    return promptTemplate;
  }

  /**
   * Get a prompt template by ID
   */
  get(id: string): PromptTemplate | undefined {
    return this.templates.get(id);
  }

  /**
   * Get a prompt template by name
   */
  getByName(name: string): PromptTemplate | undefined {
    for (const template of this.templates.values()) {
      if (template.name === name) {
        return template;
      }
    }
    return undefined;
  }

  /**
   * List all templates
   */
  list(filter?: { tags?: string[]; targetModel?: string }): PromptTemplate[] {
    let templates = Array.from(this.templates.values());

    if (filter?.tags && filter.tags.length > 0) {
      templates = templates.filter((t) =>
        filter.tags!.some((tag) => t.tags?.includes(tag)),
      );
    }

    if (filter?.targetModel) {
      templates = templates.filter((t) => t.targetModel === filter.targetModel);
    }

    return templates;
  }

  /**
   * Update a prompt template (creates new version)
   */
  update(
    id: string,
    template: string,
    changeDescription?: string,
  ): PromptTemplate | null {
    const existing = this.templates.get(id);
    if (!existing) {
      return null;
    }

    const newVersion = this.incrementVersion(existing.version);
    const timestamp = now();

    const updated: PromptTemplate = {
      ...existing,
      template,
      variables: this.extractVariables(template),
      version: newVersion,
      updatedAt: timestamp,
    };

    this.templates.set(id, updated);

    // Add version history
    const history = this.versions.get(id) || [];
    history.push({
      version: newVersion,
      template,
      createdAt: timestamp,
      changeDescription,
    });
    this.versions.set(id, history);

    return updated;
  }

  /**
   * Get version history for a template
   */
  getVersionHistory(id: string): PromptVersion[] {
    return this.versions.get(id) || [];
  }

  /**
   * Get a specific version of a template
   */
  getVersion(id: string, version: string): PromptVersion | undefined {
    const history = this.versions.get(id);
    return history?.find((v) => v.version === version);
  }

  /**
   * Rollback to a previous version
   */
  rollback(id: string, version: string): PromptTemplate | null {
    const targetVersion = this.getVersion(id, version);
    if (!targetVersion) {
      return null;
    }

    return this.update(
      id,
      targetVersion.template,
      `Rollback to version ${version}`,
    );
  }

  /**
   * Delete a template
   */
  delete(id: string): boolean {
    const deleted = this.templates.delete(id);
    this.versions.delete(id);
    return deleted;
  }

  /**
   * Render a template with variables
   */
  render(id: string, variables: Record<string, string>): string | null {
    const template = this.templates.get(id);
    if (!template) {
      return null;
    }

    return this.renderTemplate(template.template, variables);
  }

  /**
   * Render a template string directly
   */
  renderTemplate(template: string, variables: Record<string, string>): string {
    let result = template;

    for (const [key, value] of Object.entries(variables)) {
      const pattern = new RegExp(`\\{\\{\\s*${key}\\s*\\}\\}`, "g");
      result = result.replace(pattern, value);
    }

    return result;
  }

  /**
   * Compare two versions of a template
   */
  diff(
    id: string,
    version1: string,
    version2: string,
  ): {
    added: string[];
    removed: string[];
    unchanged: string[];
  } | null {
    const v1 = this.getVersion(id, version1);
    const v2 = this.getVersion(id, version2);

    if (!v1 || !v2) {
      return null;
    }

    const lines1 = v1.template.split("\n");
    const lines2 = v2.template.split("\n");
    const set1 = new Set(lines1);
    const set2 = new Set(lines2);

    return {
      added: lines2.filter((line) => !set1.has(line)),
      removed: lines1.filter((line) => !set2.has(line)),
      unchanged: lines1.filter((line) => set2.has(line)),
    };
  }

  /**
   * Export all templates
   */
  export(): Record<string, PromptTemplate> {
    const result: Record<string, PromptTemplate> = {};
    for (const [id, template] of this.templates) {
      result[id] = template;
    }
    return result;
  }

  /**
   * Import templates
   */
  import(templates: Record<string, PromptTemplate>): void {
    for (const [id, template] of Object.entries(templates)) {
      this.templates.set(id, template);
      this.versions.set(id, [
        {
          version: template.version,
          template: template.template,
          createdAt: template.createdAt,
        },
      ]);
    }
  }

  private extractVariables(template: string): string[] {
    const pattern = /\{\{\s*(\w+)\s*\}\}/g;
    const variables: Set<string> = new Set();
    let match;

    while ((match = pattern.exec(template)) !== null) {
      variables.add(match[1]);
    }

    return Array.from(variables);
  }

  private incrementVersion(version: string): string {
    const parts = version.split(".").map(Number);
    parts[2] = (parts[2] || 0) + 1;
    return parts.join(".");
  }
}
